From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: widberg <jared.widberg@zetier.com>
Date: Thu, 31 Jul 2025 14:08:47 -0400
Subject: [PATCH] 8403: Fix ghidra-stubs mypy Python type stubs problems

[ghidra-stubs] Handle void return type

[ghidra-stubs] Replace illegal characters in identifiers

[ghidra-stubs] Add this to ghidra_builtins

[ghidra-stubs] Fix unclosed Javadoc tags
---
 .../format/som/SomDynamicLoaderHeader.java    |  2 +-
 .../bin/format/som/SomExportEntryExt.java     |  2 +-
 .../typestubs/GhidraBuiltinsBuilder.java      |  3 +++
 .../typestubs/PythonTypeStubElement.java      |  2 +-
 .../typestubs/PythonTypeStubMethod.java       | 20 ++++++++++---------
 5 files changed, 17 insertions(+), 12 deletions(-)

diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomDynamicLoaderHeader.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomDynamicLoaderHeader.java
index cac53a08b7..65d596a58d 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomDynamicLoaderHeader.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomDynamicLoaderHeader.java
@@ -452,7 +452,7 @@ public class SomDynamicLoaderHeader implements StructConverter {
 	}
 
 	/**
-	 * {@return the {@link List} of {@link SomExportEntryExt export entry extensions}
+	 * {@return the {@link List} of {@link SomExportEntryExt export entry extensions}}
 	 */
 	public List<SomExportEntryExt> getExportExtensions() {
 		return exportExtensions;
diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomExportEntryExt.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomExportEntryExt.java
index 75baef79fe..fc2626879d 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomExportEntryExt.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/som/SomExportEntryExt.java
@@ -53,7 +53,7 @@ public class SomExportEntryExt implements StructConverter {
 	}
 
 	/**
-	 * {@return the size of the export symbol and is only valid for exports of type {@code ST_DATA}
+	 * {@return the size of the export symbol and is only valid for exports of type {@code ST_DATA}}
 	 */
 	public int getSize() {
 		return size;
diff --git a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/GhidraBuiltinsBuilder.java b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/GhidraBuiltinsBuilder.java
index 0a82e6e732..01eaab3463 100644
--- a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/GhidraBuiltinsBuilder.java
+++ b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/GhidraBuiltinsBuilder.java
@@ -133,6 +133,9 @@ class GhidraBuiltinsBuilder {
 			script.printField(field, printer, INDENT, false);
 			exports.add('"' + field.getSimpleName().toString() + '"');
 		}
+		printer.print(INDENT);
+		printer.println("this: GhidraScript");
+		exports.add("\"this\"");
 	}
 
 	/**
diff --git a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubElement.java b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubElement.java
index 283b454291..a98a2d8a7b 100644
--- a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubElement.java
+++ b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubElement.java
@@ -168,7 +168,7 @@ abstract class PythonTypeStubElement<T extends Element> {
 		if (PY_KEYWORDS.contains(value)) {
 			return value + "_";
 		}
-		return value;
+		return value.replace('$', '_');
 	}
 
 	/**
diff --git a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubMethod.java b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubMethod.java
index 62b6a08828..2bd5d33175 100644
--- a/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubMethod.java
+++ b/GhidraBuild/BuildFiles/Doclets/src/main/java/ghidra/doclets/typestubs/PythonTypeStubMethod.java
@@ -225,6 +225,10 @@ final class PythonTypeStubMethod extends PythonTypeStubElement<ExecutableElement
 			};
 		}
 
+		if (type.getKind() == TypeKind.VOID) {
+			return "None";
+		}
+
 		if (type instanceof DeclaredType dt) {
 			Element element = dt.asElement();
 			if (element instanceof QualifiedNameable nameable) {
@@ -314,15 +318,13 @@ final class PythonTypeStubMethod extends PythonTypeStubElement<ExecutableElement
 		printer.print(")");
 
 		TypeMirror res = el.getReturnType();
-		if (res.getKind() != TypeKind.VOID) {
-			printer.print(" -> ");
-			String convertedType = convertResultType(res);
-			if (convertedType != null) {
-				printer.print(convertedType);
-			}
-			else {
-				printer.print(sanitizeQualifiedName(res));
-			}
+		printer.print(" -> ");
+		String convertedType = convertResultType(res);
+		if (convertedType != null) {
+			printer.print(convertedType);
+		}
+		else {
+			printer.print(sanitizeQualifiedName(res));
 		}
 	}
 
-- 
2.45.1

